### DISCUSSION TITLE
TextBox StackOverflows with DateTime and IValueConverter bind.

### DISCUSSION THREAD

**Discussion by JaoHundred** [AvaloniaUI/Avalonia#13027] [Q&A] [answered]
**Describe the bug**
DateTime bind with converter to string causes infinite loop and StackOverflow in TextBox controls.

**To Reproduce**

1. Create TextBox and bind(mode TwoWay or Default) it to an DateTime
2. Create an IValueConverter to show DateTime in Text property of TextBox
3. The IValueConverter comes back and forth from and to Convert and ConvertBack methods
4. StackOverflow occurs

**Expected behavior**
IValueConverter convert string to DateTime and DateTime to string(00/00/0000 format)

**Screenshots**

![image](https://github.com/AvaloniaUI/Avalonia/assets/43298295/e0a07be6-2bc1-4b08-9043-cde2a964791b)
![image](https://github.com/AvaloniaUI/Avalonia/assets/43298295/c31552ec-c7bc-4390-871c-d26487ecce5d)
![image](https://github.com/AvaloniaUI/Avalonia/assets/43298295/5050835d-50c9-491a-836c-4284d88ed9b8)
![image](https://github.com/AvaloniaUI/Avalonia/assets/43298295/2b36f93b-877d-4f2e-9ccd-001a6b3d5f0f)

**Desktop (please complete the following information):**

- OS: Windows 10 22H2
- Version 11.0.4

**Comment by timunie** [comment]
Any reason why you do not use `DateTimePicker` ? 

Anyway, please prepare a very minimal sample as I wasn't able to get into that SO you are facing. upload or link it here. Thx.

**Comment by JaoHundred** [comment]
Hi, I didn't used DateTimePicker because I didn't wanted the user to waste much time selecting the date(I need to work with dates far beyond 1800 and lower) unless there's some way to type the date and I'm unaware of that.

[AvaloniaTextBoxTest.zip](https://github.com/AvaloniaUI/Avalonia/files/12716479/AvaloniaTextBoxTest.zip)

**Comment by timunie** [comment, marked as answer]
The issue you have is that the converter is called on every key stroke (see #3754). I uses this workaround to overcome this, but this has many drawbacks. But SO is gone at least: 

```cs
public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
{
    // this will be called on every key stroke. So make sure the new date was typed completely before trying to parse. 
    // There may be better options, one has to investigate. But we cannot convert the string before user has finished typing!
    if (value is string {Length:10} dateTime && DateTime.TryParse(dateTime, out DateTime date))
        return date.ToLocalTime();

    // In case conversion was not successful, let's do nothing instead of restetting to null
    return BindingOperations.DoNothing;
}
```

However I still think using the build in control is way better. The user is not forced to use the calendar, but they can if they want to: 

![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/cfec7f4d-0a83-41a9-a954-23c55f0f05dc)

```xml
<CalendarDatePicker SelectedDate="{Binding Date}" 
                    CustomDateFormatString="dd/MM/yyyy" />
```

-------

> Note: this is not a bug imo, so I'll convert it into a Q&A.

**Comment by JaoHundred** [comment]
This solved my issue with the Converter. CalendarDatePicker was a better solution thanks Timunie.
